blockchain/chain_sync/src/sync.rs

```diff
@@ -136,306 +128,226 @@ where
-                            }
+                    }
+                    // All other network events are being ignored currently
+                    _ => (),
+                    None => break,
+                },
+                inform_head_event = fused_inform_channel.next() => match inform_head_event {
+                    Some((peer, new_head)) => {
+                        if let Err(e) = self.inform_new_head(peer.clone(), &new_head).await {
+                            warn!("failed to inform new head from peer {}", peer);
                         }
                     }
-                    i -= REQUEST_WINDOW;
-                    continue;
+                    None => break,
                 }
-            };
-            // full tipset found in storage; validate and continue
-            let curr_epoch = fts.epoch();
-            self.validate_tipset(fts).await?;
-            self.state.write().await.set_epoch(curr_epoch);
-            i -= 1;
-            continue;
+            }
         }
+    }
 
-        Ok(())
+    /// Fetches a tipset from store or network, then passes the tipset back through the channel
+    /// to inform of the new head.
+    async fn fetch_and_inform_tipset(
+        cs: Arc<ChainStore<DB>>,
+        network: SyncNetworkContext,
+        peer_id: PeerId,
+        tsk: TipsetKeys,
+        channel: Sender<(PeerId, FullTipset)>,
+    ) {
+        match Self::fetch_full_tipset(cs.as_ref(), &network, peer_id.clone(), &tsk).await {
+            Ok(fts) => {
+                channel.send((peer_id, fts)).await;
+            }
+            Err(e) => {
+                debug!("Failed to fetch full tipset from peer ({}): {}", peer_id, e);
+            }
+        }
+    }
+
+    /// Spawns a new sync worker and pushes the state to the `ChainSyncer`
+    async fn spawn_worker(&mut self, channel: Receiver<Arc<Tipset>>) -> JoinHandle<()> {
+        let state = Arc::new(RwLock::new(SyncState::default()));
+
+        // push state to managed states in Syncer.
+        self.worker_state.write().await.push(state.clone());
+        SyncWorker {
+            state,
+            beacon: self.beacon.clone(),
+            state_manager: self.state_manager.clone(),
+            chain_store: self.chain_store.clone(),
+            network: self.network.clone(),
+            genesis: self.genesis.clone(),
+            bad_blocks: self.bad_blocks.clone(),
+            peer_manager: self.peer_manager.clone(),
+        }
+        .spawn(channel)
+        .await
     }
 
     /// informs the syncer about a new potential tipset
     /// This should be called when connecting to new peers, and additionally
     /// when receiving new blocks from the network
-    pub async fn inform_new_head(&mut self, peer: PeerId, fts: &FullTipset) -> Result<(), Error> {
+    pub async fn inform_new_head(&mut self, peer: PeerId, ts: &FullTipset) -> Result<(), Error> {
```

Comment:
I noticed in the Lotus implementation they are adding the `peerId` to their `ChainExchange` pool, should there be a TODO to add this peer to the available peers for blocksync peers?

Reply:
we're doing it when the hello message is handled by the syncer. I don't see why it would need to be handled after this process